src/node_http2.cc

```diff
@@ -281,576 +462,526 @@ ssize_t Http2Session::OnCallbackPadding(size_t frameLen,
-      nghttp2_session_get_remote_window_size(s);
-  buffer[IDX_SESSION_STATE_OUTBOUND_QUEUE_SIZE] =
-      nghttp2_session_get_outbound_queue_size(s);
-  buffer[IDX_SESSION_STATE_HD_DEFLATE_DYNAMIC_TABLE_SIZE] =
-      nghttp2_session_get_hd_deflate_dynamic_table_size(s);
-  buffer[IDX_SESSION_STATE_HD_INFLATE_DYNAMIC_TABLE_SIZE] =
-      nghttp2_session_get_hd_inflate_dynamic_table_size(s);
+    if (ret < 0)
+      return ret;
+
+    total += ret;
+  }
+  // Send any data that was queued up while processing the received data.
+  SendPendingData();
+  return total;
 }
 
-void RefreshStreamState(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-#if defined(DEBUG) && DEBUG
-  CHECK_EQ(args.Length(), 2);
-  CHECK(args[0]->IsObject());
-  CHECK(args[1]->IsNumber());
-#endif
-  int32_t id = args[1]->Int32Value(env->context()).ToChecked();
-  DEBUG_HTTP2("Http2Session: refreshing stream %d state\n", id);
-  Http2Session* session;
-  ASSIGN_OR_RETURN_UNWRAP(&session, args[0].As<Object>());
-  nghttp2_session* s = session->session();
-  Nghttp2Stream* stream;
 
-  AliasedBuffer<double, v8::Float64Array>& buffer =
-      env->http2_state()->stream_state_buffer;
+inline int32_t GetFrameID(const nghttp2_frame* frame) {
+  // If this is a push promise, we want to grab the id of the promised stream
+  return (frame->hd.type == NGHTTP2_PUSH_PROMISE) ?
+      frame->push_promise.promised_stream_id :
+      frame->hd.stream_id;
+}
 
-  if ((stream = session->FindStream(id)) == nullptr) {
-    buffer[IDX_STREAM_STATE] = NGHTTP2_STREAM_STATE_IDLE;
-    buffer[IDX_STREAM_STATE_WEIGHT] =
-        buffer[IDX_STREAM_STATE_SUM_DEPENDENCY_WEIGHT] =
-        buffer[IDX_STREAM_STATE_LOCAL_CLOSE] =
-        buffer[IDX_STREAM_STATE_REMOTE_CLOSE] =
-        buffer[IDX_STREAM_STATE_LOCAL_WINDOW_SIZE] = 0;
-    return;
-  }
-  nghttp2_stream* str =
-      nghttp2_session_find_stream(s, stream->id());
 
-  if (str == nullptr) {
-    buffer[IDX_STREAM_STATE] = NGHTTP2_STREAM_STATE_IDLE;
-    buffer[IDX_STREAM_STATE_WEIGHT] =
-        buffer[IDX_STREAM_STATE_SUM_DEPENDENCY_WEIGHT] =
-        buffer[IDX_STREAM_STATE_LOCAL_CLOSE] =
-        buffer[IDX_STREAM_STATE_REMOTE_CLOSE] =
-        buffer[IDX_STREAM_STATE_LOCAL_WINDOW_SIZE] = 0;
+inline int Http2Session::OnBeginHeadersCallback(nghttp2_session* handle,
+                                                const nghttp2_frame* frame,
+                                                void* user_data) {
+  Http2Session* session = static_cast<Http2Session*>(user_data);
+  int32_t id = GetFrameID(frame);
+  DEBUG_HTTP2SESSION2(session, "beginning headers for stream %d", id);
+
+  Http2Stream* stream = session->FindStream(id);
+  if (stream == nullptr) {
+    // Creates the stream and adds it automatically to the Http2Session
+    // TODO(jasnell): This current makes the Http2Stream instance "weak"
+    //                which may be problematic. Verify.
```

Comment:
I’d say this is a good thing and drop the `TODO`. Generally, any C++ resource that doesn’t do I/O directly on its own (i.e. libuv handles) should be weak.

Reply:
Yep. After going through this more, this being weak is going to be just fine.